### ISSUE TITLE
[DISCUSSION] C360 restart file regridding memory requirements

### ISSUE THREAD

**Issue by LiamBindle** [geoschem/GCHP#83]
I thought it might be useful to note the memory requirements for regridding restart files with GCPy. I just regridded a C360 restart file and it required 392 GB of memory.

**Comment by sdeastham** [comment]
@LiamBindle - was 392 GB of memory necessary for

1. Creation of the regrid file by xESMF; or
2. The regrid operation itself?

If (1), then perhaps we could consider putting pre-generated regrid files on one of the public servers. I've done this at MIT already for a large number of expensive regrids that I need to do somewhat frequently, since I found (1) to be expensive. If (2), we could perhaps refactor the order in which regridding is done? Although if it takes 392 GB just to regrid one layer of one species, this may need some serious thought.

**Comment by LiamBindle** [comment]
That was with `python -m gcpy.file_regrid` which uses xESMF under the hood. I'm not sure if it was the regrid operation itself or the generation of the weights, but that was the "max memory" for the job, and I thought it was worth documenting here.

The 392 GB was for a full restart file (all the species). I don't think it's too alarming because those that want to run C360 likely have access to big HPC resources, and the output restart file is ~120 GB anyways.

**Comment by WilliamDowns** [comment]
Thanks for documenting this Liam. I'll take a look at whether this peaks during the regridder file creation phase or the actual regridding.

**Comment by sdeastham** [comment]
The problem there is that 392 GB is more than might be found on a single node, even for some HPC systems - for example, Harvard's shiny new Cannon cluster has 670 nodes, but each node only has 192 GB. That means they could run a C360 simulation (easily), but couldn't actually regrid the input file unless they figured out how to make xESMF weight generation and/or regridding work in MPI (which might be trivial? But I have no idea..)

**Comment by LiamBindle** [comment]
Ah, I see what you mean. Yes that's fair&mdash;I agree, we should consider it an issue.

Some workaround that come to mind in the interim:
- request a high memory node (that's what I've been doing)
- split restart file into blocks of species (e.g., 5 groups), regrid those files individually, and then merge the resulting files

**Comment by WilliamDowns** [comment]
Regarding storing weights for regridding, we do so already [on the FTP site](http://ftp.as.harvard.edu/gcgrid/data/ExtData/GCHP/RegriddingWeights/) for a few lower resolution conversions. I don't see why we wouldn't be able to expand these offerings, especially for covering a range of non-stretched-grid regridders. Feel free to chime in right now if you know how large each regridder file is for a c360 sim.

**Comment by LiamBindle** [comment]
The total size of the weight files is on the order of a few MB for C48->C360. I suspect it's the actual regridding that's eating up all the memory

**Comment by sdeastham** [comment]
@LiamBindle - the second solution would certainly work (or could even just be handled internally in the routine I guess?) but only if the issue is with the regrid and not the weight generation, of course! I'm looking forward to @WilliamDowns  analysis though. My anecdotal findings have been that weight generation can be a memory monster (not sure why).
@WilliamDowns - ah, fantastic! It would be great to discuss a naming convention for these at some point, especially as we consider handling stretched grids and non-GEOS-Chem grids (I'd like to make sure that whatever generic convention is used is also used by gcgridobj).

**Comment by WilliamDowns** [comment]
So some tests on Cannon yielded the following peak memory consumption during different stages of `file_regrid` for an initial c48 GCHP restart file:

c90: 2GB during weight file creation, 20GB during regridding
c180: 3gb, 85gb
c360: 12gb, ran out of memory at 256gb

So the regridding itself is dominating the resource consumption. I'll look into separating into species-based chunks and other possible methods of avoiding this massive memory usage.

**Comment by LiamBindle** [comment]
Nice work! Have you look at chunking with Dask? That might be an avenue worth exploring. This PR looks like it might be helpful: https://github.com/JiaweiZhuang/xESMF/pull/26

**Comment by sdeastham** [comment]
Wow! Agreed that Dask might be part of a good solution in that case, and that (generation time aside) this deprioritizes central storage of weight files (although I'd still like to figure out a robust naming convention).

**Comment by WilliamDowns** [comment]
So it turns out the regridding was not actually the largest consumer of memory (though it does usually consume the most time); the concatenation of the 6 `xarray.Dataset` objects corresponding to each cubed-sphere face is actually where memory usage peaks at the values I wrote above. Luckily, saving each Dataset to disk individually and using `xarray.open_mfdataset()` (which makes some use of Dask under the hood, though I'm not sure exactly how) to open the 6 files as one concatenated Dataset greatly decreases memory consumption. I've pushed these changes to GCPy `dev/1.0.1`. In writing this post, I realized I've only made these changes for cubed-sphere -> cubed-sphere regridding (not lat/lon <-> cubed-sphere). I'll have to test memory consumption in taxing examples of those other regridding scenarios as well.

Peak memory consumption for regridding from c48 (these peaks now occur during regridding):

c90: 6GB
c180: 19GB
c360: 97GB
